File: sdk-rust/src/crypto.rs
//! AES-256-GCM encrypt/decrypt — format `nonce(12) || ciphertext || tag(16)`.

use aes_gcm::{
    Aes256Gcm, Nonce,
    aead::{Aead, AeadCore, KeyInit, OsRng},
};

use crate::error::HarpocrateError;

const NONCE_LEN: usize = 12;
const TAG_LEN: usize = 16;

/// Chiffre `plaintext` avec AES-256-GCM. Retourne `nonce || ciphertext+tag`.
pub fn aes_gcm_encrypt(plaintext: &[u8], key: &[u8]) -> Result<Vec<u8>, HarpocrateError> {
    if key.len() != 32 {
        return Err(HarpocrateError::Decryption(format!(
            "key must be 32 bytes, got {}",
            key.len()
        )));
    }
    let cipher = Aes256Gcm::new_from_slice(key)
        .map_err(|e| HarpocrateError::Decryption(format!("init: {e}")))?;
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);
    let ciphertext = cipher
        .encrypt(&nonce, plaintext)
        .map_err(|e| HarpocrateError::Decryption(format!("encrypt: {e}")))?;
    let mut out = Vec::with_capacity(NONCE_LEN + ciphertext.len());
    out.extend_from_slice(&nonce);
    out.extend_from_slice(&ciphertext);
    Ok(out)
}

/// Déchiffre un blob `nonce(12) || ciphertext || tag(16)` avec AES-256-GCM.
pub fn aes_gcm_decrypt(blob: &[u8], key: &[u8]) -> Result<Vec<u8>, HarpocrateError> {
    if key.len() != 32 {
        return Err(HarpocrateError::Decryption(format!(
            "key must be 32 bytes, got {}",
            key.len()
        )));
    }
    if blob.len() < NONCE_LEN + TAG_LEN {
        return Err(HarpocrateError::Decryption(format!(
            "blob too short: {} bytes",
            blob.len()
        )));
    }
    let cipher = Aes256Gcm::new_from_slice(key)
        .map_err(|e| HarpocrateError::Decryption(format!("init: {e}")))?;
    let nonce = Nonce::from_slice(&blob[..NONCE_LEN]);
    cipher
        .decrypt(nonce, &blob[NONCE_LEN..])
        .map_err(|e| HarpocrateError::Decryption(format!("invalid tag or wrong key: {e}")))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_round_trip() {
        let key = [42u8; 32];
        let plain = b"hello vault";
        let blob = aes_gcm_encrypt(plain, &key).unwrap();
        let decrypted = aes_gcm_decrypt(&blob, &key).unwrap();
        assert_eq!(decrypted, plain);
    }

    #[test]
    fn test_decrypt_with_wrong_key_fails() {
        let blob = aes_gcm_encrypt(b"x", &[1u8; 32]).unwrap();
        let err = aes_gcm_decrypt(&blob, &[2u8; 32]).unwrap_err();
        assert!(matches!(err, HarpocrateError::Decryption(_)));
    }

    #[test]
    fn test_rejects_short_key() {
        let err = aes_gcm_encrypt(b"x", &[0u8; 16]).unwrap_err();
        assert!(matches!(err, HarpocrateError::Decryption(_)));
    }

    #[test]
    fn test_rejects_short_blob() {
        let err = aes_gcm_decrypt(&[0u8; 5], &[0u8; 32]).unwrap_err();
        assert!(matches!(err, HarpocrateError::Decryption(_)));
    }
}

File: sdk-go/harpocrate/crypto.go
package harpocrate

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"
)

const (
	nonceLen = 12
	tagLen   = 16
)

// AESGCMEncrypt chiffre `plaintext` avec AES-256-GCM et retourne
// `nonce(12) || ciphertext || tag(16)`.
func AESGCMEncrypt(plaintext, key []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, &VaultDecryptionError{Reason: fmt.Sprintf("key must be 32 bytes, got %d", len(key))}
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, &VaultDecryptionError{Reason: "aes init: " + err.Error()}
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, &VaultDecryptionError{Reason: "gcm init: " + err.Error()}
	}
	nonce := make([]byte, nonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return nil, &VaultDecryptionError{Reason: "nonce: " + err.Error()}
	}
	ct := aead.Seal(nil, nonce, plaintext, nil)
	out := make([]byte, 0, nonceLen+len(ct))
	out = append(out, nonce...)
	out = append(out, ct...)
	return out, nil
}

// AESGCMDecrypt déchiffre un blob `nonce(12) || ciphertext || tag(16)`.
func AESGCMDecrypt(blob, key []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, &VaultDecryptionError{Reason: fmt.Sprintf("key must be 32 bytes, got %d", len(key))}
	}
	if len(blob) < nonceLen+tagLen {
		return nil, &VaultDecryptionError{Reason: fmt.Sprintf("blob too short: %d bytes", len(blob))}
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, &VaultDecryptionError{Reason: "aes init: " + err.Error()}
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, &VaultDecryptionError{Reason: "gcm init: " + err.Error()}
	}
	plain, err := aead.Open(nil, blob[:nonceLen], blob[nonceLen:], nil)
	if err != nil {
		return nil, &VaultDecryptionError{Reason: "invalid tag or wrong key: " + err.Error()}
	}
	return plain, nil
}
